File: frontend/src/types/job.ts
export interface JobStep {
  id: string;
  step_number: number;
  operation_type: string;
  operation_params: Record<string, unknown>;
  result_image_path: string | null;
  processing_time_ms: number | null;
  status: "pending" | "processing" | "completed" | "failed";
  created_at: string;
}

export interface Job {
  id: string;
  user_id: string;
  pipeline_id: string | null;
  image_id: string;
  status: "pending" | "processing" | "completed" | "failed";
  current_step: number;
  total_steps: number;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  steps?: JobStep[];
}

export interface JobProgress {
  job_id: string;
  step?: number;
  total?: number;
  operation?: string;
  result_path?: string;
  status: "processing" | "completed" | "failed";
  error?: string;
}

File: backend/app/models/job.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class JobStatus(str, enum.Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class Job(Base):
    __tablename__ = "jobs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"))
    pipeline_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("pipelines.id", ondelete="SET NULL"), nullable=True)
    image_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("images.id", ondelete="CASCADE"))
    status: Mapped[JobStatus] = mapped_column(Enum(JobStatus), default=JobStatus.PENDING)
    current_step: Mapped[int] = mapped_column(Integer, default=0)
    total_steps: Mapped[int] = mapped_column(Integer)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    user = relationship("User", back_populates="jobs")
    pipeline = relationship("Pipeline", back_populates="jobs")
    steps = relationship("JobStep", back_populates="job", cascade="all, delete-orphan", order_by="JobStep.step_number")
